maco/collector.py

```diff
@@ -98,133 +70,38 @@ def __init__(
-                # Assume that because it's the closest, it's the most relevant
-                return os.path.join(parent_dir, "venv")
-            else:
-                # Keep searching in the parent directory for a venv
-                return find_venv(parent_dir)
-
-        # walk packages in the extractors directory to find all extactors
-        extractors = {}
-        for module_path, module_name, ispkg in pkgutil.walk_packages(mod.__path__, mod.__name__ + "."):
-            if ispkg:
-                # skip __init__.py
-                continue
-            if not module_name.endswith(filename):
-                # if filename was specified, skip modules that don't have that name
-                continue
-            logger.debug(f"inspecting '{module_name}' for extractors")
-            # raise an exception if one of the potential extractors can't be imported
-            # note that excluding an extractor through include/exclude does not prevent it being imported
-
-            # Local site packages, if any, need to be loaded before attempting to import the module
-            parser_venv = find_venv(module_path.path)
-            parser_site_packages = None
-            if parser_venv:
-                for dir in glob(os.path.join(parser_venv, "lib/python*/site-packages")):
-                    sys.path.insert(1, dir)
-                    break
-            try:
-                module = importlib.import_module(module_name)
-            except Exception as e:
-                # Log if there was an error importing module
-                logger.error(f"{module_name}: {e}")
-                continue
-            finally:
-                if parser_site_packages in sys.path:
-                    sys.path.remove(parser_site_packages)
-
-            # find extractors in the module
-            for _, member in inspect.getmembers(module):
-                if not inspect.isclass(member):
-                    # not a class
-                    continue
-                if not issubclass(member, extractor.Extractor):
-                    # not an extractor
-                    continue
-                if issubclass(member, extractor.Extractor) and not member.author:
-                    # not a valid extractor
-                    continue
+            # Recursively create/update virtual environments
+            utils.create_venv(path_extractors, logger=logger)
 
+        def extractor_module_callback(member, module, venv) -> bool:
+            if utils.maco_extractor_validation(member):
                 # check if we want this extractor
                 name = member.__name__
                 if self.exclude and name in self.exclude:
                     logger.debug(f"exclude excluded '{name}'")
-                    continue
+                    return
                 if self.include and name not in self.include:
                     logger.debug(f"include excluded '{name}'")
-                    continue
+                    return
```

Comment:
this is the same as exclude, should this return a "positive response" such as True?

Reply:
Actually, should both the exclude and include checks be done before the maco_extractor_validation call?